Allow additional async redraw even if the first redraw is skipped

kfule opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This PR allows asynchronous redraw processing even if the first redraw is skipped by setting `event.redraw=false` before await in the async function.
Also, this PR avoids redraw after the dom or component is removed.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This PR is intended to address the following cases of [zulip](https://mithril.zulipchat.com/#narrow/channel/324076-general/topic/Should.20async.20event.20handlers.20be.20a.20semver.20major.20change.3F/with/516572123)

> If it was set to `false` before the first await and then `true` after it, the first redraw should be skipped, but the second one shouldn't.

In #3020 , I didn't consider the case where `event.redraw` would later be `true`.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
`npm run test` with additional test

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] My change requires a documentation update, and I've opened a pull request to update it already:
- [x] I have read https://mithril.js.org/contributing.html.

dead-claudia reviewed (changes requested):
I missed this during the first review, but while we're at it, can you make sure to skip redraws after the component is removed, so removals don't trigger further redraws? (This bug also exists with sync event handlers, being able to fire events post-removal, and we should fix both in a single step.)

You should be able to get away with just doing this:

1. Add a `if (self._ != null)` check to the condition in the resolver handler where `self` is the outer event listener object.
2. Add an `if (vnode.events != null) vnode.events._ = null` right before [this line](https://github.com/MithrilJS/mithril.js/blob/c105f134d2ab06d52dce0e9d103eec341f40c6de/render/render.js#L629).

That'd fix both.

kfule commented:
@dead-claudia Thanks for your review.
As you suggested, I have changed the code so that it does not redraw after removing a dom or component (with the tests.)
With this commit, I removed the `eventRedraw` variable that holds `this._`, because `self` holds `this`.

kfule commented:
@dead-claudia 
This is not directly related to this PR, but it appears that the number of assertions in [`Test and maybe release / run-tests / test-node`](https://github.com/MithrilJS/mithril.js/actions/runs/14931181879/job/41947598107?pr=3021#step:5:30) is the number of `main` branch (16789 assertions) and not the number of branches in this PR (just 16800 assertions).
The appropriate branch may not have been checked out in the test Actions.

dead-claudia commented:
> @dead-claudia 
> 
> This is not directly related to this PR, but it appears that the number of assertions in [`Test and maybe release / run-tests / test-node`](https://github.com/MithrilJS/mithril.js/actions/runs/14931181879/job/41947598107?pr=3021#step:5:30) is the number of `main` branch (16789 assertions) and not the number of branches in this PR (just 16800 assertions).
> 
> The appropriate branch may not have been checked out in the test Actions.

That's a good catch. Filed https://github.com/MithrilJS/infra/issues/69 to track it down. (This is in a common workflow script.)

dead-claudia commented on render/tests/test-event.js:
```diff
@@ -884,4 +916,48 @@ o.spec("event", function() {
 			})
 		})
 	})
+	o("avoid sync redraw after removal", function() {
```
Can you extend this test to also verify the event listener itself wasn't called?

That way, we know it couldn't possibly attempt an async redraw later.

kfule replied:
I added assertions about whether the event listener itself was called or not.

The event listener itself is not removed and will be called, but it should be clearer that redraw is not being called.

dead-claudia commented on render/tests/test-event.js:
```diff
@@ -884,4 +916,48 @@ o.spec("event", function() {
+		render(root, div)
+		div.dom.dispatchEvent(e)
+
+		o(redraw.callCount).equals(1)
+		o(redraw.this).equals(undefined)
+		o(redraw.args.length).equals(0)
```
Can you also assert `thenCB` is set? That being set also implies that the event listener was actually called.

kfule replied:
OK, added the following assertions.
- Whether the event listener itself has been called or not.
- Whether thenCB was set or not.

dead-claudia commented on render/tests/test-event.js:
```diff
@@ -848,6 +848,38 @@ o.spec("event", function() {
+		render(root, div)
+		div.dom.dispatchEvent(e)
+
+		o(redraw.callCount).equals(0)
```
Can you also assert `thenCB` is set? That being set also implies that the event listener was actually called.

kfule replied:
ditto

dead-claudia reviewed (changes requested):
See comments.

Should hopefully be the last request for changes.

kfule commented:
@dead-claudia Thanks for your kind review.
I added assertions about whether the event listeners themselves were called or not where redraw would not be called.
